lib/Zonemaster/CLI.pm

```diff
@@ -407,6 +407,77 @@ sub run {
+        my @existing_test_modules = keys %existing_tests;
+        my @existing_test_cases = map { @{ $existing_tests{$_} } } @existing_test_modules;
+
+        foreach my $t ( @{ $self->test } ) {
+            # There should be at most one slash character
+            if ( $t =~ tr/\/// > 1 ) {
+                die __( "Error: Invalid input '$t' in --test. There must be at most one slash ('/') character.\n");
+            }
+
+            # The case does not matter
+            $t = lc( $t );
+
+            my ( $module, $method );
+            # Fully qualified module and test case (e.g. Example/example12), or just a test case (e.g. example12). Note the different capturing order.
+            if ( ( ($module, $method) = $t =~ m#^ ( [a-z]+ ) / ( [a-z]+[0-9]{2} ) $#ix )
+                or
+                 ( ($method, $module) = $t =~ m#^ ( ( [a-z]+ ) [0-9]{2} ) $#ix ) )
+            {
+                # Check that test module exists
+                if ( grep( /^$module$/,  map { lc($_) } @existing_test_modules ) ) {
+                    # Check that test case exists
+                    if ( grep( /^$method$/, @existing_test_cases ) ) {
+                        if ( not grep( /^$method$/, @{ Zonemaster::Engine::Profile->effective->get( 'test_cases' ) } ) ) {
+                            # Actual forcing will be done later in the code
+                            say $fh_diag __x( "Notice: Engine does not have test case '$method' enabled in the profile. Forcing...");
+                        }
+
+                        push @testing_suite, "$module/$method";
+                    }
+                    else {
+                        die __( "Error: Unrecognized test case '$method' in --test. Use --list-tests for a list of valid choices.\n" );
+                    }
+                }
+                else {
+                    die __( "Error: Unrecognized test module '$module' in --test. Use --list-tests for a list of valid choices.\n" );
+                }
+            }
+            # Just a module name (e.g. Example) or something invalid.
+            else {
+                $t =~ s{/$}{};
+                # Check that test module exists
+                if ( grep( /^$t$/,  map { lc($_) } @existing_test_modules ) ) {
+                    push @testing_suite, $t;
+                }
+                else {
+                    die __( "Error: Invalid input '$t' in --test.\n" );
+                }
+            }
+        }
+
+        my @actual_test_cases;
+        foreach my $t ( @testing_suite ) {
+            # Either a module/method, or just a module
+            my ( $module, $method ) = split('/', $t);
+            if ( $method ) {
+                push @actual_test_cases, $method;
+            }
+            else {
+                # Get the test module with the right case
+                ( $module ) = grep { lc( $module ) eq lc( $_ ) } @existing_test_modules;
+                push @actual_test_cases, @{ $existing_tests{$module} };
```

Comment:
In current develop if you specify `--profile file --test dnssec` and the given profile contains a strict subset of all dnssec test cases, only that subset will be run. With this implementation all dnssec test cases will be run. I believe you agree the behavior in current develop is more reasonable. So after expanding the module name to a list of test case names but before pushing onto @actual_test_cases we should grep the list to only keep those test cases that are currently included in the test_cases profile property. I'm sorry I wasn't clear on this last time around.

Reply:
Oh, you already spotted it. I didn't see your comment before.